### PR TITLE
Add: `take` function and getitem for lazyarray

### PR DESCRIPTION
**Pull request by yangguohao**
Hi @jcmgray Johnnie. I have implemented both of the `autoray.lazy.take` and fancy indexing of `autoray.lazy.getiem`

For `autoray.lazy.take`, I have considered following two situations
1.  Indices may be `list` or `tuple`,
2.  The indices may be LazyArray, or indices should be backend-type array. For instance, the indices should be `torch.Tensor` for `torch.take`.

For `autoray.lazy.getitem`, the indices can only be the `list` or `tuple`, and raise the exception when the indices is wrong for LazyArray.

I'm not sure that I covered all the complicated cases for these two functions, these all can be further improved

### CODE REVIEW

File: autoray/lazy/core.py [hunk 1]
```diff
@@ -1267,6 +1267,43 @@ def where(condition, x, y):
...
+        if isinstance(sequence, (list, tuple)):
+            if all(isinstance(x, int) for x in sequence):
+                shape_list.append(len(sequence))
+            elif all((isinstance(x, (list, tuple)) and len(x) == len(sequence[0])) for x in sequence):
+                shape_list.append(len(sequence))
+                helper(sequence[0])
+            else:
+                raise ValueError(
+                    'setting an array element with a sequence. '
+                )
```
Comment (jcmgray) [inline autoray/lazy/core.py:1283]: Some comments might be helpful here as recursive logic is always a little hard to follow.

**Review by jcmgray — CHANGES_REQUESTED** [review]
Thanks for the PR! The main thing missing is unit tests, especially with a function with this many use-cases.

I'd write a `pytest.mark.parametrize`, and for all the combinations of relevant indices, check that the shape matches the `numpy` call with the equivalent args or errors.

File: autoray/lazy/core.py (hunk 1 above)
Comment (yangguohao) [reply autoray/lazy/core.py:1283]: Sure, I will add the comment soon.

Comment (yangguohao) [reply autoray/lazy/core.py]: Instead of using the recursive function, I used a new method to get the shape of indices, it is much more simple.

**Comment by yangguohao** [comment]
Add the unit tests.

File: autoray/lazy/core.py [hunk 2]
```diff
@@ -1267,6 +1267,30 @@ def where(condition, x, y):
     )
 
 
+def indices_shape_for_lazy(indices):
+    import numpy as np
+    shape_list = list(np.array(indices).shape)
+    return shape_list
```
Comment (jcmgray) [inline autoray/lazy/core.py:1273]: Would a function like so suffice here? :
```python
def _get_py_shape(x):
    """Infer the shape of a possibly nested list/tuple object.
    """
    if hasattr(x, "shape"):
        return x.shape
    if isinstance(x, (tuple, list)):
        return (len(x),) + _get_py_shape(x[0])
    return ()
```
then we could avoid the `numpy` dependency.

**Comment by jcmgray** [comment]
Great, thanks for the PR and changes!

**Comment by jcmgray** [comment]
I think we might have broken some behavior here, e.g. `x[:, 0]` now gives `shape=()`. Probably needed quite a lot more test coverage of getitem, since it has many different variants.

**Comment by yangguohao** [comment]
> I think we might have broken some behavior here, e.g. `x[:, 0]` now gives `shape=()`. Probably needed quite a lot more test coverage of getitem, since it has many different variants.

Seems like I forgot to consider the `...` and `:` type index. I will go check and fix it.

**Comment by jcmgray** [comment]
I actually did already hear: https://github.com/jcmgray/autoray/commit/aa4955a2fd9c835a1184de1a04eddea948929f83
